src/rpc/rawtransaction.cpp

```diff
@@ -1284,6 +1310,126 @@ static std::string WriteHDKeypath(std::vector<uint32_t>& keypath)
     return keypath_str;
 }
 
+UniValue blindpsbt(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            RPCHelpMan{"blindpsbt",
+                "\nUses the blinding data from the PSBT inputs to generate the blinding data for the PSBT outputs.\n",
+                {
+                    {"psbt", RPCArg::Type::STR, RPCArg::Optional::NO, "The PSBT base64 string"},
+                    {"ignoreblindfail", RPCArg::Type::BOOL, /* default*/ "true", "Return a transaction even when a blinding attempt fails due to number of blinded inputs/outputs."},
+                },
+                RPCResult{
+                    "\"psbt\"       (string) The base64-encoded partially signed transaction\n"
+                },
+                RPCExamples{
+                    HelpExampleCli("blindpsbt", "\"psbt\"")
+                    + HelpExampleRpc("blindpsbt", "\"psbt\"")
+                }
+            }.ToString()
+        );
+
+    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VBOOL}, true);
+
+    // Unserialize the transactions
+    PartiallySignedTransaction psbtx;
+    std::string error;
+    if (!DecodeBase64PSBT(psbtx, request.params[0].get_str(), error)) {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("TX decode failed %s", error));
+    }
+
+    bool fIgnoreBlindFail = true;
+    if (!request.params[1].isNull()) {
+        fIgnoreBlindFail = request.params[1].get_bool();
+    }
+
+    // TODO(gwillen): Refactor out significant duplicated code between here and rawblindrawtransaction.
+
+    std::vector<CAmount> input_amounts;
+    std::vector<uint256> input_blinds;
+    std::vector<uint256> input_asset_blinds;
+    std::vector<CAsset> input_assets;
+    std::vector<uint256> output_value_blinds;
+    std::vector<uint256> output_asset_blinds;
+    std::vector<CAsset> output_assets;
+    std::vector<CPubKey> output_pubkeys;
+
+    int n_blinded_ins = 0;
+
+    // TODO(gwillen): If blinding is not possible due to missing input data, we should bail here with a useful error message.
+    for (const auto& input : psbtx.inputs) {
+        input_blinds.push_back(input.value_blinding_factor);
+        input_asset_blinds.push_back(input.asset_blinding_factor);
+        input_assets.push_back(input.asset);
+        input_amounts.push_back(input.value ? *input.value : CAmount(-1));
+
+        if (!input_blinds.back().IsNull()) {
+            n_blinded_ins++;
+        }
+    }
+
+    for (auto& output : psbtx.outputs) {
+        output_pubkeys.push_back(output.blinding_pubkey);
+    }
+
+    // How many are we trying to blind?
+    int num_pubkeys = 0;
+    unsigned int keyIndex = -1;
```

Comment:
Flagged by ubsan for being an implicit signedness conversion (which is not UB but ubsan lints for it). But also this is literally UB because keyIndex is never checked against this -1 value, it's just used as an array index.

Reply:
Should be fixed by #900, just noting here for posterity.

Reply:
Oh, this block appears to have been copied from rawblindrawtransaction which has had this issue since at least https://github.com/ElementsProject/elements/pull/515

Reply:
Oh, my bad, in both cases it is actually safe becasue `keyIndex` is only -1 if `num_pubkeys` is left at 0, and `keyIndex` is only used as an index when `num_pubkeys` is nonzero.